arm_isa.cpp

```diff
@@ -1377,7 +1377,7 @@ inline void LDRH(arm_isa* ref, int rd, int rn,
     printf("Unpredictable LDRH instruction result (Address is not Halfword Aligned)\n");
     return;
   }
-  value = MEM.read(ref->ls_address.entire);
+  value = MEM.read_half(ref->ls_address.entire);
   value &= 0xFFFF; /* Zero extends halfword value 
 		      BUG: Model must be little endian in order to the code work  */
```

Comment:
I removed the 0xFFFF mask and the bug came back (!). So: 
(1) using 'read' and '0xFFFF' mask results in 'Seg Fault'; 
(2) using 'read_half' without '0xFFFF' results in 'Seg Fault'; 
(3) using 'read_half' with '0xFFFF' results success. 

To reproduce the bug, create a model with cache:

arm_cache.ac:
AC_ARCH(arm){
  ac_mem     MEM:512M;
  ac_regbank RB:16;

  ac_icache   IC("2w", 64, 8, "wt", "random");
  ac_dcache   DC("2w", 64, 8, "wt", "random");

  // ac_tlm_intr_port inta;  // system level interrupt port
  ac_reg R14_irq, R14_fiq, R14_svc, R14_abt, R14_und, R13_irq, R13_svc;
  ac_reg R13_abt, R13_und, R13_fiq;
  ac_reg SPSR_irq, SPSR_fiq, SPSR_svc, SPSR_abt, SPSR_und;
  // FIQ private regs
  ac_reg R12_fiq, R11_fiq, R10_fiq, R9_fiq, R8_fiq;
  ac_wordsize 32;
  ac_reg id;  
  ARCH_CTOR(arm) {
    ac_isa("arm_isa.ac");
    set_endian("little");
    IC.bindTo(MEM);
    DC.bindTo(MEM);
  };
};

Generate the simulator:
acsim arm_cache.ac

Run the hello world
./arm.x --load=/tmp/hello.arm